src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs

```diff
@@ -17,101 +17,150 @@ public static partial class JsonSerializer
         private static void ReadCore(
             JsonSerializerOptions options,
             ref Utf8JsonReader reader,
-            ref ReadStack state)
+            ref ReadStack readStack)
         {
             try
             {
-                while (reader.Read())
+                JsonReaderState initialState = default;
+
+                while (true)
                 {
+                    if (options.ReadAhead)
+                    {
+                        // When we're reading ahead we always have to save the state
+                        // as we don't know if the next token is an opening object or
+                        // array brace.
+                        initialState = reader.CurrentState;
+                    }
+
+                    if (!reader.Read())
+                    {
+                        // Need more data
+                        break;
+                    }
+
                     JsonTokenType tokenType = reader.TokenType;
 
                     if (JsonHelpers.IsInRangeInclusive(tokenType, JsonTokenType.String, JsonTokenType.False))
                     {
                         Debug.Assert(tokenType == JsonTokenType.String || tokenType == JsonTokenType.Number || tokenType == JsonTokenType.True || tokenType == JsonTokenType.False);
 
-                        if (HandleValue(tokenType, options, ref reader, ref state))
-                        {
-                            continue;
-                        }
+                        HandleValue(tokenType, options, ref reader, ref readStack);
                     }
                     else if (tokenType == JsonTokenType.PropertyName)
                     {
-                        HandlePropertyName(options, ref reader, ref state);
+                        HandlePropertyName(options, ref reader, ref readStack);
                     }
                     else if (tokenType == JsonTokenType.StartObject)
                     {
-                        if (state.Current.SkipProperty)
+                        if (readStack.Current.SkipProperty)
                         {
-                            state.Push();
-                            state.Current.Drain = true;
+                            readStack.Push();
+                            readStack.Current.Drain = true;
                         }
-                        else if (state.Current.IsProcessingValue)
+                        else if (readStack.Current.IsProcessingValue)
                         {
-                            if (HandleValue(tokenType, options, ref reader, ref state))
+                            if (!HandleObjectAsValue(tokenType, options, ref reader, ref readStack, ref initialState))
```

Comment:
How do we know we are in "object" case here?

Reply:
Because we're in the JsonTokenType..StartObject case.